hazelcast/src/main/java/com/hazelcast/config/AtomicLongConfig.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.config;
+
+/**
+ * Contains the configuration for an {@link java.util.concurrent.atomic.AtomicLong}.
+ */
+public class AtomicLongConfig extends AbstractBasicConfig<AtomicLongConfig> {
+
+    private transient AtomicLongConfigReadOnly readOnly;
+
+    public AtomicLongConfig() {
+    }
+
+    public AtomicLongConfig(String name) {
+        setName(name);
+    }
+
+    public AtomicLongConfig(AtomicLongConfig config) {
+        super(config);
+    }
+
+    @Override
+    public String toString() {
+        return "AtomicLongConfig{"
+                + fieldsToString()
+                + "}";
+    }
+
+    @Override
+    public int getId() {
+        return ConfigDataSerializerHook.ATOMIC_LONG_CONFIG;
+    }
+
+    /**
+     * Gets immutable version of this configuration.
+     *
+     * @return immutable version of this configuration
+     * @deprecated this method will be removed in 4.0; it is meant for internal usage only
+     */
+    @Override
+    public AtomicLongConfig getAsReadOnly() {
```

Comment:
Could it be that there is a threading issue here due to the lack of a happens before relation?

I'm not sure which thread can call getAsReadonly and if this can be done more than once.

Reply:
We use exactly this pattern in several config classes. If this is an issue, it's not a new one. But I think the `getAsReadOnly()` is idempotent, so in the worst case we'll create some garbage.

The method is used in this context:
```java
public class AtomicLongService implements ManagedService, RemoteService,
    MigrationAwareService {

    private final ConstructorFunction<String, AtomicLongContainer>
        atomicLongConstructorFunction
            = new ConstructorFunction<String, AtomicLongContainer>() {
                public AtomicLongContainer createNew(String key) {
                   // findAtomicLongConfig() returns the read-only config <------
                    AtomicLongConfig config = nodeEngine.getConfig()
                        .findAtomicLongConfig(key);
                    return new AtomicLongContainer(config);
                }
            };

    public AtomicLongContainer getLongContainer(String name) {
        return getOrPutIfAbsent(containers, name, atomicLongConstructorFunction);
    }
}
```

Reply:
Ha.. It isn't a matter of idempotent. It could be that the second thread sees a partially initialized object. So it could be that is sees that the AtomicLongConfig reference, but perhaps it doesn't see that backup-count is 2.

Reply:
I guess it is only called from the appropriate partition thread and therefor you automatically get the program order happens before rule.